index.js

```diff
@@ -63,34 +128,137 @@ module.exports = () => {
 
     selectIpfsReady: state => state.ipfs.ready,
 
-    selectIpfsApiAddress: state => state.ipfs.apiAddress,
+    selectIpfsProvider: state => state.ipfs.provider,
+
+    selectIpfsApiOpts: state => state.ipfs.apiOpts,
 
-    selectIpfsInitFailed: state => !!state.ipfs.error,
+    selectIpfsApiAddress: state => `/ip4/${state.ipfs.apiOpts.host}/tcp/${state.ipfs.apiOpts.port}`,
+
+    selectIpfsInitFailed: state => state.ipfs.failed,
 
     selectIpfsIdentity: state => state.ipfs.identity,
 
-    doInitIpfs: () => async ({ dispatch, getState }) => {
-      dispatch({ type: 'IPFS_INIT_STARTED' })
+    doInitIpfs: () => async (store) => {
+      await getIpfs(opts, store)
+    },
 
-      const { apiAddress } = getState().ipfs
-      let identity = null
+    doStopIpfs: () => async ({ dispatch }) => {
+      ipfs.stop(() => {
+        dispatch({ type: 'IPFS_STOPPED' })
+      })
+    },
 
-      try {
-        ipfs = await getIpfs({ api: true, ipfs: apiAddress })
-        // will fail if remote api is not available on default port
-        identity = await ipfs.id()
-        // if it works, save the address for the next time
-        persistAddress(apiAddress)
-      } catch (error) {
-        return dispatch({ type: 'IPFS_INIT_FAILED', error })
-      }
+    doUpdateIpfsApiOpts: (usrOpts) => (store) => {
+      saveUserOpts('ipfsApi', usrOpts)
+      store.dispatch({ type: 'IPFS_API_OPTS_UPDATED', payload: usrOpts })
 
-      dispatch({ type: 'IPFS_INIT_FINISHED', payload: identity })
+      getIpfs(Object.assign({}, opts, {
+        tryWindow: false,
+        tryJsIpfs: false
+      }), store)
     },
 
-    doUpdateIpfsAPIAddress: (apiAddress) => ({dispatch, store}) => {
-      dispatch({type: 'IPFS_API_UPDATED', payload: apiAddress})
-      store.doInitIpfs()
+    doUpdateIpfsApiAddress: (addr) => ({ store }) => {
+      addr = addr.split('/')
+
+      store.doUpdateIpfsApiOpts({
+        host: addr[2],
+        port: addr[4],
+        protocol: 'http'
```

Comment:
We need to check the multi addr to see if it specifies `https` rather than assuming `http`